### DISCUSSION TITLE
Incorrect multi-table `DELETE..USING` SQL query when deleting Joined Inheritance Relationship object declared using `__mapper_args__ = {'with_polymorphic': '*'}` with SQAlchemy 1.4

### DISCUSSION THREAD

**Discussion by artaod** [sqlalchemy/sqlalchemy#6178] [Usage Questions]
**Describe the bug**
`Query.delete()` generates incorrect multi-table `DELETE..USING` SQL query when deleting parent object declared by mapper with `__mapper_args__ = {'with_polymorphic': '*'}` in Joined Inheritance Relationship.

Incorrect SQL query:
```sql
sqlalchemy.exc.ProgrammingError: (mysql.connector.errors.ProgrammingError) 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'LEFT OUTER JOIN engineer ON employee.id = engineer.id LEFT OUTER JOIN manager ON' at line 1
[SQL:
DELETE FROM employee
LEFT OUTER JOIN engineer ON employee.id = engineer.id
LEFT OUTER JOIN manager ON employee.id = manager.id
USING employee
LEFT OUTER JOIN engineer ON employee.id = engineer.id
LEFT OUTER JOIN manager ON employee.id = manager.id, engineer, manager
WHERE employee.name = %(name_1)s]
[parameters: {'name_1': 'test'}]
```

**Expected behavior**
Expected to get SQL statement rather like this:
```sql
DELETE FROM employee, engineer, manager
USING employee
LEFT OUTER JOIN engineer ON employee.id = engineer.id
LEFT OUTER JOIN manager ON employee.id = manager.id
WHERE employee.name = %(name_1)s]
```

**To Reproduce**
1. Define Relationships with Joined Inheritance as shown here 
https://docs.sqlalchemy.org/en/14/orm/inheritance.html#relationships-with-joined-inheritance
2. Add `'with_polymorphic': '*'` to `__mapper_args__` of parent class as shown here https://docs.sqlalchemy.org/en/14/orm/inheritance_loading.html?highlight=with_polymorphic#setting-with-polymorphic-at-mapper-configuration-time

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session

engine = create_engine("mysql+mysqlconnector://test@localhost/test", echo=True)

Base = declarative_base()


class Employee(Base):
	__tablename__ = 'employee'
	id = Column(Integer, primary_key=True)
	name = Column(String(50))
	type = Column(String(50))

	__mapper_args__ = {
		'polymorphic_identity': 'employee',
		'polymorphic_on': type,
		'with_polymorphic': '*'  # Adding this causes incorrect "DELETE..USING" SQL
	}


class Engineer(Employee):
	__tablename__ = 'engineer'
	id = Column(Integer, ForeignKey('employee.id'), primary_key=True)
	engineer_name = Column(String(30))

	__mapper_args__ = {
		'polymorphic_identity': 'engineer',
	}


class Manager(Employee):
	__tablename__ = 'manager'
	id = Column(Integer, ForeignKey('employee.id'), primary_key=True)
	manager_name = Column(String(30))

	__mapper_args__ = {
		'polymorphic_identity': 'manager',
	}


Base.metadata.create_all(engine)
session = Session(engine)

session.query(Employee).filter(Employee.name == 'test').delete()

```

**Error**

```
INFO sqlalchemy.engine.Engine BEGIN (implicit)
INFO sqlalchemy.engine.Engine DELETE FROM employee LEFT OUTER JOIN engineer ON employee.id = engineer.id LEFT OUTER JOIN manager ON employee.id = manager.id USING employee LEFT OUTER JOIN engineer ON employee.id = engineer.id LEFT OUTER JOIN manager ON employee.id = manager.id, engineer, manager WHERE employee.name = %(name_1)s
INFO sqlalchemy.engine.Engine [generated in 0.00043s] {'name_1': 'test'}
Traceback (most recent call last):
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\sqlalchemy\engine\base.py", line 1705, in _execute_context
    self.dialect.do_execute(
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\sqlalchemy\engine\default.py", line 681, in do_execute
    cursor.execute(statement, parameters)
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\mysql\connector\cursor.py", line 551, in execute
    self._handle_result(self._connection.cmd_query(stmt))
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\mysql\connector\connection.py", line 490, in cmd_query
    result = self._handle_result(self._send_cmd(ServerCmd.QUERY, query))
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\mysql\connector\connection.py", line 395, in _handle_result
    raise errors.get_exception(packet)
mysql.connector.errors.ProgrammingError: 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'LEFT OUTER JOIN engineer ON employee.id = engineer.id LEFT OUTER JOIN manager ON' at line 1

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\User\py\test_delete_with_polymorphic.py", line 48, in <module>
    session.query(Employee).filter(Employee.name=='test').delete()
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\sqlalchemy\orm\query.py", line 3087, in delete
    result = self.session.execute(
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\sqlalchemy\orm\session.py", line 1662, in execute
    result = conn._execute_20(statement, params or {}, execution_options)
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\sqlalchemy\engine\base.py", line 1520, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\sqlalchemy\sql\elements.py", line 313, in _execute_on_connection
    return connection._execute_clauseelement(
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\sqlalchemy\engine\base.py", line 1389, in _execute_clauseelement
    ret = self._execute_context(
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\sqlalchemy\engine\base.py", line 1748, in _execute_context
    self._handle_dbapi_exception(
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\sqlalchemy\engine\base.py", line 1929, in _handle_dbapi_exception
    util.raise_(
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\sqlalchemy\util\compat.py", line 211, in raise_
    raise exception
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\sqlalchemy\engine\base.py", line 1705, in _execute_context
    self.dialect.do_execute(
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\sqlalchemy\engine\default.py", line 681, in do_execute
    cursor.execute(statement, parameters)
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\mysql\connector\cursor.py", line 551, in execute
    self._handle_result(self._connection.cmd_query(stmt))
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\mysql\connector\connection.py", line 490, in cmd_query
    result = self._handle_result(self._send_cmd(ServerCmd.QUERY, query))
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\mysql\connector\connection.py", line 395, in _handle_result
    raise errors.get_exception(packet)
sqlalchemy.exc.ProgrammingError: (mysql.connector.errors.ProgrammingError) 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'LEFT OUTER JOIN engineer ON employee.id = engineer.id LEFT OUTER JOIN manager ON' at line 1
[SQL: DELETE FROM employee LEFT OUTER JOIN engineer ON employee.id = engineer.id LEFT OUTER JOIN manager ON employee.id = manager.id USING employee LEFT OUTER JOIN engineer ON employee.id = engineer.id LEFT OUTER JOIN manager ON employee.id = manager.id, engineer, manager WHERE employee.name = %(name_1)s]
[parameters: {'name_1': 'test'}]
(Background on this error at: http://sqlalche.me/e/14/f405)
```

**Versions.**
 - OS: Windows 8.1
 - Python: 3.9.2
 - SQLAlchemy: 1.4.3
 - Database: MariaDB 10.2.14
 - DBAPI: mysql_connector 2.2.9

**Additional context**
SQLAlchemy 1.3 has no such error as it produces single-table DELETE statement.

Thank you for your attention.

**Have a nice day!**

**Comment by zzzeek** [comment]
hey there -

MySQL's DELETE..USING syntax is fairly specific to MySQL, while Postrgresql I think might also supports some variant of this, these structures are not generally available in SQL on most other databases, and the ORM generally can't support such consttructs.

For this reason, ORM-enabled UPDATE and DELETE of multiple rows are explicitly not supported for joined-table inheritance setups, and this is documented in the big pink dragon box at https://docs.sqlalchemy.org/en/14/orm/session_basics.html#orm-expression-update-delete :

> ORM-enabled UPDATEs and DELETEs do not handle joined table inheritance automatically. If the operation is against multiple tables, typically individual UPDATE / DELETE statements against the individual tables should be used. Some databases support multiple table UPDATEs. Similar guidelines as those detailed at Multiple Table Updates may be applied.

in short, none of this is tested or supported and it's out of scope; if you need mass UPDATE/DELETE of these structures, invoke your SQL using session.connection().execute() so that no ORM functionalities are used.

**Reply by artaod** [reply]
Thank you for quick reply!

Actually I don't need exactly mass-delete in my case, but just don't crash with error when deleting the parent records as it was in SQLAlchemy 1.3 (descendant object deletion can be performed by previous statement at application level).

Is there any option available to have `Query.delete()` don't go "multi-delete" and fire only single-table DELETE statement while having `__mapper_args__ = {'with_polymorphic': '*'}` in place? Or a relatively easy possibility to add such an option?

**Reply by zzzeek** [reply]
Use the Table objects directly:

```python
stmt = delete(Engineer.__table__).where(Engineer.id == Employee.id).where(Employee.name == 'test')
session.connection().execute(stmt)

stmt = delete(Manager.__table__).where(Manager.id == Employee.id).where(Manager.name == 'test')
session.connection().execute(stmt)

stmt = delete(Employee.__table__).where(Employee.name == 'test')
session.connection().execute(stmt)

```

**Reply by zzzeek** [reply]
SQLAlchemy also doesnt support the MySQL DELETE..USING variant where there are multiple target tables, we can only render additional FROMs.    no other database supports this concept.

the most portable way to delete from all the tables is to use ON DELETE CASCADE on the foreign keys, then just DELETE from the `Employee.__table__ ` in all cases.

**Reply by artaod** [reply]
Then maybe it would be reasonable to force `Query.delete()` to don't even try to produce `DELETE..USING` for multiple tables and delete just one? To prevent SQL syntax error happening.

**Reply by zzzeek** [reply]
I would have to research all the possibilities here for what kinds of mapped tables are and are not appropriate for DELETE on what drivers, it's not clear what works and what doesn't.    We usually have to do something more sophisticated than "if with_polymorphic is not None: raise", mappings can be against any kinds of selectables which may or may not work.    The database itself is the best authority on what SQL works and what doesn't so in the absense of a really well thought out plan for everything that will definitely not work and everything that might work, we just let the database do it.

**Comment by christiansicari** [comment]
I have the same issue here, but in my case I don't have specified the `with_polymorphic`.
This my structure
```python
class SuperClass(BaseORMModel):
__mapper_args__ = {
        "polymorphic_identity": "component",
        "polymorphic_on": type,
    }
...
class Subclass(SuperClass):

   ....
    id = Column(UUID(as_uuid=True), ForeignKey("superclass.id", ondelete="CASCADE"), primary_key=True)
    text = Column(TEXT)

    __mapper_args__ = {"polymorphic_identity": "text", "polymorphic_load": "inline"}
```
This query
```python
stmt = delete(SuperClass).where(SuperClass.id == class_id)
delete_result = self.db.execute(stmt)
````
causes something like
```
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.SyntaxError) syntax error at or near "LEFT"
E       LINE 1: DELETE FROM superclass LEFT OUTER JOIN subclass O...
```

**Reply by christiansicari** [reply]
In my case the issue came "polymorphic_load": "inline", changing with "polymorphic_load": "selectin", seems to solved the problem

**Reply by zzzeek** [reply]
delete from `Cls.__table__` to just remove the ORM from the equation
